Fixing environment detection behavior on GAE Managed VMs.

theacodes opened the pull request:
When running on Managed VMs, the evironment variables are set for "SERVER_SOFTWARE" but the user doesn't necessarily have to be using a appengine-compat runtime such as `google/appengine-python`, they could be using a general purpose one such as `google/python-runtime`. Because the generic one doesn't have the app engine sdk, it would crash when trying to obtain default credentials. This fixes the problem by checking if the app engine sdk is importable.

nathanielmanistaatgoogle commented on tests/test_oauth2client.py:
```diff
@@ -200,12 +201,20 @@ def test_create_scoped(self):
+    sys.modules['google'] = object()
+    sys.modules['google.appengine'] = object()
     os.environ['SERVER_SOFTWARE'] = 'Google App Engine/XYZ'
     self.assertEqual('GAE_PRODUCTION', _get_environment())
```
If this assertion fails, won't the sys.modules space be polluted by this test and possibly interfere with other tests?

theacodes replied:
Possibly, and it could lead to making it difficult to determine which test is causing this. I can re-work this as a context manager if desired.

nathanielmanistaatgoogle replied:
It is strongly desired.

nathanielmanistaatgoogle commented on tests/test_oauth2client.py:
```diff
@@ -200,12 +201,20 @@ def test_create_scoped(self):
+    sys.modules['google'] = object()
+    sys.modules['google.appengine'] = object()
     os.environ['SERVER_SOFTWARE'] = 'Development/XYZ'
     self.assertEqual('GAE_LOCAL', _get_environment())
```
Same question here.

jonparrott commented:
@nathanielmanistaatgoogle context manager added.

nathanielmanistaatgoogle commented on tests/test_oauth2client.py:
```diff
@@ -144,6 +146,23 @@ def get(self, key, default=None):
+    path = '.'.join(parts[:n+1])
+    sys.modules[path] = object()
+  
+  try:
```
I don't understand why yield is wrapped in a try/finally?

theacodes replied:
To prevent exceptions from causing leaks into sys.modules. See [here](https://docs.python.org/2/library/contextlib.html#contextlib.closing)

nathanielmanistaatgoogle replied:
Wow, today I learned that until now I've been writing dangerously defective code with contextlib.contextmanager. :-( Thanks for the education!

theacodes replied:
Haha!
![marvin](http://i2.kym-cdn.com/photos/images/original/000/847/015/01a.jpg)

nathanielmanistaatgoogle commented on tests/test_oauth2client.py:
```diff
@@ -144,6 +146,23 @@ def get(self, key, default=None):
+    yield
+
+  finally:
+    for n in range(len(parts)):
```
In the first half of the function create a sequence of the dummy entries added so that you don't have to recompute them to remove them here in the second half.

nathanielmanistaatgoogle commented on tests/test_oauth2client.py:
```diff
@@ -144,6 +146,23 @@ def get(self, key, default=None):
+def mock_module_import(module):
+  """Place a dummy object in sys.modules to mock an import test"""
+  parts = module.split('.')
+  for n in range(len(parts)):
```
If I say that I prefer "i" for an index and "n" to represent a quantity, am I being too fussy?

theacodes replied:
I'll accept it. :smile:

nathanielmanistaatgoogle commented:
Please squash commits with each round of code review back-and-forth; the successive stages of progress here shouldn't necessarily become a part of canonical project history.

Commit:
Fixing environment detection behavior on GAE Managed VMs.

    When running on Managed VMs, the evironment variables are set for "SERVER_SOFTWARE" but the user doesn't necessarily have to be using a appengine-compat runtime such as ``google/appengine-python``, they could be using a general purpose one such as ``google/python-runtime``. Because the generic one doesn't have the app engine sdk, it would crash when trying to obtain default credentials. This fixes the problem by checking if the app engine sdk is importable.

jonparrott commented:
Commit squashed and the context manager was updated with your feedback.

nathanielmanistaatgoogle commented on tests/test_oauth2client.py:
```diff
@@ -144,6 +146,22 @@ def get(self, key, default=None):
+  """Place a dummy object in sys.modules to mock an import test"""
+  parts = module.split('.')
+  entries = ['.'.join(parts[:i+1]) for i in range(len(parts))]
+  for n in entries:
```
Please don't use "n" for values that aren't integers.

I think "entry" or "fully_qualified_module_name" are both fine names for this value.

nathanielmanistaatgoogle commented on tests/test_oauth2client.py:
```diff
@@ -144,6 +146,22 @@ def get(self, key, default=None):
 
+@contextlib.contextmanager
+def mock_module_import(module):
+  """Place a dummy object in sys.modules to mock an import test"""
```
Nitpick: missing period at end of text.

jonparrott commented:
Got the nits taken care of.

nathanielmanistaatgoogle commented:
Looks good to me. @craigcitro or @soltanmm, please be a second reviewer on this?

anthmgoogle commented:
LGTM.

nathanielmanistaatgoogle commented:
Thank you @anthmgoogle.

jonparrott commented:
Thanks, everyone!
